### ISSUE TITLE
Petitboot

### ISSUE THREAD

**Issue by ptr1337** [erdii/odroid-hc4-arch#1]
Hey dude, 

im trying since long time to get the odroid hc4 to work with archlinuxarm and aarch64. 

Manjaro first alpha unstable release worked with petitboot, after the "stable" release and also the lasts updates it didnt worked anymore. 

However, did you managed to get it booted from petitboot ? 

regards

**Comment by erdii** [comment]
Hey @ptr1337 :wave:
Nope, I've only used u-boot so far. Shameless self-plug: u-boot build process [1]

That said, I'd love to get rid of u-boot and use petitboot but I had a really hard time searching for petitboot / petitboot+odroid-hc4 documentation.
Do you have any pointers, example configuration files, links to documentation?

Cheers! :)

[1] https://github.com/erdii/odroid-hc4-uboot-build

**Comment by ptr1337** [comment]
You can test the manjaro alpha unstable iso, 

https://mirror.fkardame.com/Linux/Images/Odroid/HC4/

there is also the stable iso which dont work with petitboot anymore. 
i just handled it know that im holding the kernel and its running perfectly with ssd + 4tb hdd.

The performance is also much higher then using ubuntu 21.04 with the netinstaller, dont know why. 

i also tried to compile a kernel with the same config, but newer commit but this failed last time at me. 

just change in the pacman.conf to ignorepkgs = linux-vim linux-vim-headers and then update it normally and it works flawless.

**Comment by ptr1337** [comment]
https://forum.odroid.com/viewtopic.php?f=203&t=42059

some discussions here

**Comment by erdii** [comment]
Thanks :)

FWIW archlinuxarm's `linux-aarch64-rc` has a 5.12 kernel with mainline devicetree for the hc4; provided by @tobetter who works at hardkernel. (Big praise! :raised_hands:)

**Comment by ptr1337** [comment]
Could be a try worth. 

I'll compile it and share it with you.

**Comment by tobetter** [comment]
Hey guys,
I've tested the image, https://mirror.fkardame.com/Linux/Images/Odroid/HC4/Manjaro-ARM-mate-hc4-21.03-alpha-unstable.img.xz, on HC4 and it boots. I've not completed all setup, at least Petitboot know what to boot up with the image.

Is there any other image that does not boot or can you share the version of the Petitboot, where the image is flashed, please?

**Comment by ptr1337** [comment]
No, only the aplpha unstable image boots. if updating including the kernel it wont boot anymore. 

Test the "stable" image from him and it wont boot from ssd/hdd anymore. 

i also tested to update to the newst kernel 5.13.1 and also not booted.

If i ignore the kernel in the pacman.conf (linux-vim linux-vim-heders) it will update completly and also stuff like uboot and so on. 

Mostly ill stay on the old kernel since it feels for me much faster then the ubuntu image and im mostly a complete arch user. 

 I tested really much, as you can read in the thread. but i didnt got it to boot. after the kernel 5.11.4

**Comment by ptr1337** [comment]
@tobetter 

download smthh from there, and it wont boot. 

https://mirror.fkardame.com/Linux/Images/Odroid/HC4/21.04/

**Comment by tobetter** [comment]
@ptr1337 Is the problem that Manjaro does not start/get into desktop after setup or Petitboot does not show the boot menu for the Manjaro. I am stuck at Manjaro logo before logging session. Testing with KDE image.

EDIT: I've updated kernel to 5.13.0 with pacman and it boots by the Petitboot. Is there anything I am missing?

**_[    2.794402] kvm: pmu event creation failed -2
[    2.857494] panfrost ffe40000.gpu: dev_pm_opp_set_regulators: no regulator (mali)9
[    3.505382] dw-pcie fc000000.pcie: IRQ index 1 not found
Starting version 248.3-2-arch
[    9.167445] meson-pcie fc000000.pcie: invalid resource
[    9.261454] rtc-pcf8563 0-0051: pcf8563_write_block_data: err=-6 addr=0e, data=03
[    9.263831] rtc-pcf8563 0-0051: pcf8563_probe: write error
[    9.319630] meson-pcie fc000000.pcie: error: wait linkup timeout

Manjaro Linux 5.13.0-1-MANJARO-ARM (ttyAML0)

linuxfactory login: tobetter
Password: 
Welcome to Manjaro-ARM
~~Website: https://manjaro.org
~~Forum:   https://forum.manjaro.org/c/arm
~~Matrix:  #manjaro-arm-public:matrix.org
Last login: Wed Jul 21 19:59:54 on ttyAML0
[tobetter@linuxfactory ~]$ uname -a
Linux linuxfactory.or.kr 5.13.0-1-MANJARO-ARM #1 SMP PREEMPT Wed Jun 30 11:12:14 UTCx
[tobetter@linuxfactory ~]$_**

**Comment by ptr1337** [comment]
Did you updated it competly after it ? 

sudo pacman -Syu

for me it didnt worked with the latest kernel from manjaro.

**Comment by ptr1337** [comment]
@ptr1337 Is the problem that Manjaro does not start/get into desktop after setup or Petitboot does not show the boot menu for the Manjaro. I am stuck at Manjaro logo before logging session. Testing with KDE image.

this i got also when i tested with montior

**Comment by ptr1337** [comment]
Total Download Size:   50.59 MiB
Total Installed Size:  76.73 MiB
Net Upgrade Size:       1.44 MiB

:: Proceed with installation? [Y/n] y
:: Retrieving packages...
 linux-vim-5.13.0-1-aarch64                50.6 MiB  73.3 MiB/s 00:01 [#######################################] 100%
(1/1) checking keys in keyring                                        [#######################################] 100%
(1/1) checking package integrity                                      [#######################################] 100%
(1/1) loading package files                                           [#######################################] 100%
(1/1) checking for file conflicts                                     [#######################################] 100%
(1/1) checking available disk space                                   [#######################################] 100%
:: Processing package changes...
(1/1) upgrading linux-vim                                             [#######################################] 100%
:: Running post-transaction hooks...
(1/5) Arming ConditionNeedsUpdate...
(2/5) Updating module dependencies...
(3/5) Updating linux-vim module dependencies...
(4/5) Updating linux-vim initcpios...
==> Building image from preset: /etc/mkinitcpio.d/linux-vim.preset: 'default'
  -> -k 5.13.0-1-MANJARO-ARM -c /etc/mkinitcpio.conf -g /boot/initramfs-linux.img
==> Starting build: 5.13.0-1-MANJARO-ARM
  -> Running build hook: [base]
  -> Running build hook: [udev]
  -> Running build hook: [plymouth]
  -> Running build hook: [autodetect]
  -> Running build hook: [modconf]
  -> Running build hook: [block]
  -> Running build hook: [filesystems]
  -> Running build hook: [keyboard]
  -> Running build hook: [fsck]
==> Generating module dependencies
==> Creating gzip-compressed initcpio image: /boot/initramfs-linux.img
==> Image generation successful
(5/5) Updating uimg initcpios...
Image Name:   Ramdisk Image
Created:      Wed Jul 21 23:09:02 2021
Image Type:   AArch64 Linux RAMDisk Image (uncompressed)
Data Size:    7784856 Bytes = 7602.40 KiB = 7.42 MiB
Load Address: 00000000
Entry Point:  00000000
:: Looking for AUR upgrades
:: Looking for devel upgrades
 there is nothing to do
reboot

and dead again:

(5/5) Updating uimg initcpios...
Image Name:   Ramdisk Image
Created:      Wed Jul 21 23:09:02 2021
Image Type:   AArch64 Linux RAMDisk Image (uncompressed)
Data Size:    7784856 Bytes = 7602.40 KiB = 7.42 MiB
Load Address: 00000000
Entry Point:  00000000
:: Looking for AUR upgrades
:: Looking for devel upgrades
 there is nothing to do
╭─       ptr1337@arch    ~                                  24.14s    2.33G    0.07    23:09:05   
╰─ sudo reboot               
Connection to 192.168.188.25 closed by remote host.
Connection to 192.168.188.25 closed.
ptr1337@CachyOS ~ [255]> ssh ptr1337@192.168.188.25
ssh: connect to host 192.168.188.25 port 22: No route to host

**Comment by tobetter** [comment]
> @ptr1337 Is the problem that Manjaro does not start/get into desktop after setup or Petitboot does not show the boot menu for the Manjaro. I am stuck at Manjaro logo before logging session. Testing with KDE image.
> 
> this i got also when i tested with montior

I believe this is the Petitboot. This happened to me after **_sudo pacman -Syu_** and rebooting. It's stuck somewhere while booting and it works after rebooting and had no issue multiple times after rebooting.

> :: Looking for devel upgrades
> there is nothing to do
> reboot
> 
> and dead again:
> 
> (5/5) Updating uimg initcpios...

This is not clear to me if your ODROID-HC4 had stops and you had rebooted and updating is proceed, or it's stooped for some time and moving forward to next phase. This does look like updating has been stopped while ODROID-HC4 is alive.

> Connection to 192.168.188.25 closed by remote host.
> Connection to 192.168.188.25 closed.
> ptr1337@CachyOS ~ [255]> ssh ptr1337@192.168.188.25
> ssh: connect to host 192.168.188.25 port 22: No route to host

I wonder if you had seen the Petitboot and see the boot entry with "NO LABEL" for Manjaro. I've tested multiple ODROID-HC4 with the same Manjaro image multiple times, all boots without failure. FYI, the Petitboot only involved to load OS files, kernel/ramdisk/device tree, and it does not resident once Kernel starts.

**Comment by tobetter** [comment]
> Hey @ptr1337 
> Nope, I've only used u-boot so far. Shameless self-plug: u-boot build process [1]
> 
> That said, I'd love to get rid of u-boot and use petitboot but I had a really hard time searching for petitboot / petitboot+odroid-hc4 documentation.
> Do you have any pointers, example configuration files, links to documentation?
> 
> Cheers! :)
> 
> [1] https://github.com/erdii/odroid-hc4-uboot-build

Can you let me know what problem you have with the Petitboot? The Manjaro image I've tested can boot from the Petitboot, the Petitboot is basically uses the boot script same as U-Boot, some commands in the boot script would not compatible but the images I've tested works.

**Comment by ptr1337** [comment]
The images which I use from the net installer and your work never faced such problems. Anyway to reset the petitboot ? 

I also use a ssd for booting also tested a second ssd. 

Yes when I boot I see in the petit boot the no label and stuck in the Plymouth screen

**Comment by erdii** [comment]
> > Hey @ptr1337
> > Nope, I've only used u-boot so far. Shameless self-plug: u-boot build process [1]
> > That said, I'd love to get rid of u-boot and use petitboot but I had a really hard time searching for petitboot / petitboot+odroid-hc4 documentation.
> > Do you have any pointers, example configuration files, links to documentation?
> > Cheers! :)
> > [1] https://github.com/erdii/odroid-hc4-uboot-build
> 
> Can you let me know what problem you have with the Petitboot? The Manjaro image I've tested can boot from the Petitboot, the Petitboot is basically uses the boot script same as U-Boot, some commands in the boot script would not compatible but the images I've tested works.

Thank you for popping in @tobetter! :)

My problem is that i didn't find any documentation on how to configure petitboot - but I guess I could extract configuration from the OS-images that hardkernel provides (and links in the device wiki [1]) or the manjaro image mentioned earlier?

I'm currently using u-boot and i didn't erase petitboot from flash so it is probably "in-use" during boot, but I think (pls correct me if i'm wrong here) that I *could* get rid of u-boot and have petitboot directly boot/kexec the kernel with the right configuration?

Edit: added os-options from which i could extract petitboot config

[1] https://wiki.odroid.com/odroid-hc4/os_images/ubuntu

**Comment by erdii** [comment]
Also unrelated side-note: on archlinuxarm (with the config present in this repository) the odroid-hc4 can not reboot without a manual power-cycle. Is this a hardware or software problem?
Should I post this in the hardkernel forum?

Edit: i have a friend running a similar archlinuxarm config and they have the same problem.

**Comment by tobetter** [comment]
> > > Hey @ptr1337
> > > Nope, I've only used u-boot so far. Shameless self-plug: u-boot build process [1]
> > > That said, I'd love to get rid of u-boot and use petitboot but I had a really hard time searching for petitboot / petitboot+odroid-hc4 documentation.
> > > Do you have any pointers, example configuration files, links to documentation?
> > > Cheers! :)
> > > [1] https://github.com/erdii/odroid-hc4-uboot-build
> > 
> > 
> > Can you let me know what problem you have with the Petitboot? The Manjaro image I've tested can boot from the Petitboot, the Petitboot is basically uses the boot script same as U-Boot, some commands in the boot script would not compatible but the images I've tested works.
> 
> Thank you for popping in @tobetter! :)
> 
> My problem is that i didn't find any documentation on how to configure petitboot - but I guess I could extract configuration from the OS-images that hardkernel provides (and links in the device wiki [1]) or the manjaro image mentioned earlier?
> 
> I'm currently using u-boot and i didn't erase petitboot from flash so it is probably "in-use" during boot, but I think (pls correct me if i'm wrong here) that I _could_ get rid of u-boot and have petitboot directly boot/kexec the kernel with the right configuration?
> 
> Edit: added os-options from which i could extract petitboot config
> 
> [1] https://wiki.odroid.com/odroid-hc4/os_images/ubuntu

Petitboot use the boot script of U-boot, but currently the commands in the script can be correctly parsed and work with Hardkernel stock U-Boot, not upstream U-Boot. For some OS image has a bit different configuration and I am trying to fix it both script or the Petitboot. I've tested OS image you mentioned and it works for me. I need an OS which does not work.

Maybe you can grap the files in the BOOT partition from your OS image that does not boot with the Petitboot, the boot script, Linux kernel, device tree and initramfs, so I can test them if the Petitboot reads and can get into kernel.

**Comment by erdii** [comment]
I'm pretty sure that i got it working now! :)
Its essentially these files here https://github.com/erdii/odroid-hc4-arch/tree/main/root/boot + kernel image + devicetree files + Kernel uimg.

I made a tarball of a working odroid's boot partition (without spiboot because the file is deleted on first boot): 
[odroid-hc4-boot-partition.tar.gz](https://github.com/erdii/odroid-hc4-arch/files/6887461/odroid-hc4-boot-partition.tar.gz)

What I am now interested in is:
- am i allowed to redistribute spiboot.img in this repo?
- how to build petitboot spiboot.img completely from source because i took that file from the hardkernel ubuntu image?
- how to determine petitboot/u-boot boot script values like devicetree and kernel load-addresses?

**Comment by tobetter** [comment]
> I'm pretty sure that i got it working now! :)
> Its essentially these files here https://github.com/erdii/odroid-hc4-arch/tree/main/root/boot + kernel image + devicetree files + Kernel uimg.
> 
> I made a tarball of a working odroid's boot partition (without spiboot because the file is deleted on first boot):
> [odroid-hc4-boot-partition.tar.gz](https://github.com/erdii/odroid-hc4-arch/files/6887461/odroid-hc4-boot-partition.tar.gz)
> 
> What I am now interested in is:
> 
> * am i allowed to redistribute spiboot.img in this repo?
> * how to build petitboot spiboot.img completely from source because i took that file from the hardkernel ubuntu image?
> * how to determine petitboot/u-boot boot script values like devicetree and kernel load-addresses?

So now no more issue to boot from the Petitboot? What's changed?

For the update, run **_pb-update_** after select "Exit to shell" in the Petitboot. This will update its firmware through the network if installed firmware is outdated. You do not necessary to distribute it this will make more confustion which firmware is recent version or old version can overwrite new one, updating through the network will manage the right firmware.
Basically load addres in the boot script has no meaning in the Petitboot other than indicating which is kernel and which device tree. Even if you put wrong address for Linux kernel and make the address matches between **_load_** and **_booti/m_** command, it will boot.

**Comment by erdii** [comment]
> So now no more issue to boot from the Petitboot? What's changed?

I used the spiboot.img from hardkernel's latest ubuntu image to upgrade petitboot. :)

> For the update, run pb-update after select "Exit to shell" in the Petitboot. This will update its firmware through the network if installed firmware is outdated. You do not necessary to distribute it this will make more confustion which firmware is recent version or old version can overwrite new one, updating through the network will manage the right firmware.

Is there a way have petitboot update headless? I'd really like to avoid attaching screen+keyboard or uart.

> Basically load addres in the boot script has no meaning in the Petitboot other than indicating which is kernel and which device tree. Even if you put wrong address for Linux kernel and make the address matches between load and booti/m command, it will boot.

yeah i noticed that too and it's totally confusing - how can i verify that the devicetree actually works though?

> how to build petitboot spiboot.img completely from source because i took that file from the hardkernel ubuntu image?

I would really like to know so I can (ideally) create a pipeline that builds the petitboot image like i did with u-boot here: [1][2]

[1] https://github.com/erdii/odroid-hc4-uboot-build
[2] https://github.com/erdii/odroid-hc4-uboot-build/releases/tag/odroidg12-v2015.01-3

**Comment by ptr1337** [comment]
Did you got it sloved ?
Do you have anywhere a image for booting ?

**Comment by erdii** [comment]
I have a root filesystem tarball that I can send you - I just need to find a place to upload it to.

**Comment by ptr1337** [comment]
ehm, i got a fileserver. os petitboot working ?

https://seafile.ptr1337.dev/u/d/8359170e2e884d91b42e/

there you can upload :)

**Comment by erdii** [comment]
I think so yes - There is a small possibillity that u-boot is still in the MBR of the sd-cards that I reused but i deliberately overrode the MBR with dd.

I'm already uploading the archive to gdrive at this current time. Will post the link later.

You can also build yourself by checking out this repo and running some commands:

Prerequisites (that i currently know) are:
- arch-install-scripts
- qemu-user-static
- binfmt-qemu-static (needs to be reinstalled because package hooks fail on first installation - idk why)
- tar
- wget
- gpg

check out the repo, cd into the folder and run:

```
make .cache/ArchLinuxARM-aarch64-latest.tar.gz # download latest aarch rootfs archive
make .cache/root  # unpack archive
make setup # pre-configure system

# add you ssh keys to the root user
curl https://github.com/ptr1337.keys | sudo tee .cache/root/root/.ssh/authorized_keys
sudo chmod 0600 .cache/root/root/.ssh/authorized_keys
```

now you can prepare the sd-card:
```
sudo scripts/format.sh /dev/your/sd/card/device
sudo scripts/mount.sh /dev/your/sd/card/device
sudo scripts/sync_root.sh .cache/root .cache/mnt
sudo scripts/unmount.sh
```

**Comment by erdii** [comment]
Upload finished: https://drive.google.com/file/d/1UgfIinVFdvt-foYIQEhXRez5b7loHLXt/view?usp=sharing
this is the zip of `.cache/root` so you'd still need to format the sd-card, then extract the archive (stripping path-prefix `.cache/root/`) into the root partition and move `/boot` into the boot partition.

**Comment by ptr1337** [comment]
Thanks for the info. 

But booting from the ssd is not possible or do i understand this wrong ? :D 

sorry for the question

**Comment by erdii** [comment]
Btw. you did not remove petitboot like suggested here [1]?:

```
# flash_eraseall /dev/mtd0
# flash_eraseall /dev/mtd1
# flash_eraseall /dev/mtd2
# flash_eraseall /dev/mtd3
```

[1] https://www.armbian.com/odroid-hc4/

**Comment by erdii** [comment]
> But booting from the ssd is not possible or do i understand this wrong ? :D

i don't get the question?

if you mean: Is booting from the sd-card, created by this process, possible?
then the answer is: yes - i currently have 3 odroid-hc4 booted from such sd-cards.

if you mean: is booting from an SSD that is inserted into an SATA-slot possible?
then the answer is: i don't know - i have never tried.

**Comment by ptr1337** [comment]
Ah okay. 

I meant from booting from the SSD directly. Not with the SPI Boot. 
As i already mentioned, it was possile with the alpha version of manjaro. 
After kernel it isnt anymore.

**Comment by erdii** [comment]
> Not with the SPI Boot.

I'm not sure what you mean with that? I am under the assumption that `petitboot == SPI boot`.

Since you asked about SSD support:

I have moved one of my odroids to boot from SSD and it works fine.
No micro-sd card present in this odroid at all! :)
And it uses the latest `linux-aarch64-rc` kernel.

PS: totally unrelated - your github profile links to the url `httos://cachyos.org ` - that's probably a typo

**Comment by ptr1337** [comment]
Wow, i got it to boot with your uploaded dir. 

with the script i got at make setup everytime errors because pacman-contrib. 

The only problem which i got right now: 

`alarm@odroid-hc4-one ~]$ sudo pacman -Sy
sudo: /etc/sudo.conf is owned by uid 1000, should be 0
sudo: /usr/bin/sudo must be owned by uid 0 and have the setuid bit set
[alarm@odroid-hc4-one ~]$ id
uid=1000(alarm) gid=1000(alarm) groups=1000(alarm),998(wheel)
[alarm@odroid-hc4-one ~]$ pacman -Sy
error: you cannot perform this operation unless you are root.
[alarm@odroid-hc4-one ~]$ sudo su
sudo: /etc/sudo.conf is owned by uid 1000, should be 0
sudo: /usr/bin/sudo must be owned by uid 0 and have the setuid bit set
[alarm@odroid-hc4-one ~]$ su -
Password: 
su: cannot set groups: Operation not permitted
[alarm@odroid-hc4-one ~]$ su - root
Password: 
su: cannot set groups: Operation not permitted
`

Edit: worked now with the makefile. ill try tosetup it. :x

**Comment by ptr1337** [comment]
Worked like a charm and running. Using SSD and my 8 TB WD RED

Thanks for your work!

**Comment by ptr1337** [comment]
I just compiled right now with the linux-aarch64-rc pkgbuild the 5.14-rc7 and is running like a charm. 
i can upload it if someone is interested in.

**Comment by erdii** [comment]
Nice :) I'm ok with my current kernel but I have a question:
Does rebooting work with the newer kernel?

**Comment by ptr1337** [comment]
> Nice :) I'm ok with my current kernel but I have a question:
> Does rebooting work with the newer kernel?

Wiithout plugging out and in again ?

ill check one moment

**Comment by ptr1337** [comment]
@erdii 

Yes, rebooted without issues. 

```
[alarm@odroid-hc4-one ~]$ sudo reboot
Connection to 192.168.188.25 closed by remote host.
Connection to 192.168.188.25 closed.
~/kernel/linux-cacule-aur master ❯ ssh alarm@192.168.188.25                  ✘ 255 3m 8s
(alarm@192.168.188.25) Password:
Last login: Wed Sep  1 14:36:27 2021 from 192.168.188.29
[alarm@odroid-hc4-one ~]$
```

**Comment by ptr1337** [comment]
@erdii 

One thing if possbile, dont use this for the root a ssh key if possbile, or give the alarm user sudo/wheel permissions. 


I'll going to compile today the 5.14 stable and ill see if its boots, if then ill upload it.

regards

**Comment by breadcat** [comment]
Hi there, I've tried using the makefile then the scripts to create a SSD housed image, and I'm getting ArchLinux listed on /dev/sda1/ in PetitBoot, however when I boot from this I get nothing but a black screen over HDMI, a solid blue LED on the unit and my OLED display freezes.

Can anybody let me know what I've potentially done wrong, or potentially upload a known working image?

**Comment by ptr1337** [comment]
> Hi there, I've tried using the makefile then the scripts to create a SSD housed image, and I'm getting ArchLinux listed on /dev/sda1/ in PetitBoot, however when I boot from this I get nothing but a black screen over HDMI, a solid blue LED on the unit and my OLED display freezes.
> 
> Can anybody let me know what I've potentially done wrong, or potentially upload a known working image?

I don't got it the last time also not to work when I have tried it.
I used now the manjaro installation which is basically on arch and it did worked so far. Its "basically" also based on arch. Right now I did not tried to convert the manjaro repo's to the arch linux ones. But it should maybe work. 

I think that the kernel just need to get updated at archlinuxarm to be compatible again with the odroid hc4

**Comment by breadcat** [comment]
No worries, thanks for the reply. My first assumption was that I'd done something wrong.

I tried Manjaro but on first boot it wouldn't detect USB keyboard input which was a little less than ideal.

**Comment by ptr1337** [comment]
> No worries, thanks for the reply. My first assumption was that I'd done something wrong.
> 
> I tried Manjaro but on first boot it wouldn't detect USB keyboard input which was a little less than ideal.

I just use my odroid over ssh as a "server". So I cant really say much about the desktop usage
